File: languages/c/blur-par.c
// blur-par: parallel (wall-clock scaling track) 3x3 Gaussian blur. Same result as
// serial blur.c for ANY core count. Invocation: blur-par <cores> <n>.
// Per-pass row-band decomposition: within each of PASSES double-buffered passes,
// worker w writes output rows [w*N/cores, (w+1)*N/cores) of dst, reading the
// read-only src buffer (including neighbour rows + clamped borders). The join is
// the pass barrier; then src/dst swap. Output is bit-identical to serial.
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

static long long now_ns(void) { struct timespec t; clock_gettime(CLOCK_MONOTONIC, &t); return (long long)t.tv_sec * 1000000000LL + t.tv_nsec; }

#define P 1000000007L
#define PASSES 4

static long lcg(long s) { return (s * 1103515245L + 12345L) & 0x7fffffffL; }
static int clampi(int x, int n) { return x < 0 ? 0 : (x >= n ? n - 1 : x); }

static const int Kk[9] = {1, 2, 1, 2, 4, 2, 1, 2, 1};
static int N;
static int *src, *dst;

typedef struct { int lo, hi; } Band;

static void *worker(void *arg) {
    Band *b = (Band *)arg;
    for (int i = b->lo; i < b->hi; i++) {
        for (int j = 0; j < N; j++) {
            int acc = 0;
            for (int di = -1; di <= 1; di++) {
                int ni = clampi(i + di, N);
                for (int dj = -1; dj <= 1; dj++) {
                    int nj = clampi(j + dj, N);
                    acc += Kk[(di + 1) * 3 + (dj + 1)] * src[(long)ni * N + nj];
                }
            }
            dst[(long)i * N + j] = acc / 16;
        }
    }
    return NULL;
}

int main(int argc, char **argv) {
    int cores = argc > 1 ? atoi(argv[1]) : 1;
    N = argc > 2 ? atoi(argv[2]) : 256;
    if (cores < 1) cores = 1;
    if (cores > N) cores = N;
    if (cores > 256) cores = 256;

    src = malloc((size_t)N * N * sizeof(int));
    dst = malloc((size_t)N * N * sizeof(int));
    long s = 42;
    for (long k = 0; k < (long)N * N; k++) { s = lcg(s); src[k] = (int)(s % 256); }

    long long _t0 = now_ns();
    for (int pass = 0; pass < PASSES; pass++) {
        pthread_t th[256];
        Band bands[256];
        for (int w = 0; w < cores; w++) {
            bands[w].lo = (int)((long)w * N / cores);
            bands[w].hi = (int)((long)(w + 1) * N / cores);
            pthread_create(&th[w], NULL, worker, &bands[w]);
        }
        for (int w = 0; w < cores; w++) pthread_join(th[w], NULL);
        int *t = src; src = dst; dst = t;   // barrier passed -> swap
    }
    fprintf(stderr, "COMPUTE_NS %lld\n", now_ns() - _t0);

    long h = 0;
    for (long k = 0; k < (long)N * N; k++) h = (h * 31 + src[k]) % P;
    printf("%ld\n", h);
    printf("blur(%d)\n", N);

    free(src); free(dst);
    return 0;
}

File: languages/rust/blur-par.rs
// blur-par: parallel 2D Gaussian blur for the wall-clock scaling track.
// Invocation: blur-par <cores> <n>
// Output: identical to serial blur at the same n (core-invariant).
//
// Decomposition: per pass, partition output rows into `cores` contiguous bands.
// Each worker reads the full input buffer (including neighbour rows for the 3x3
// stencil) and writes only its output rows.  After each pass all workers join
// (implicit at scope end) before the buffers are swapped — same barrier semantics
// as the serial double-buffer loop.  Border clamping (edge-replication) is
// preserved verbatim from the serial code.
//
// Build:
//   rustc -O languages/rust/blur-par.rs -o /app/blur-par
use std::env;

const P: i64 = 1000000007;
const PASSES: usize = 4;
const K: [i32; 9] = [1, 2, 1, 2, 4, 2, 1, 2, 1];

fn lcg(s: i64) -> i64 {
    (s.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7fffffff
}

fn clampi(x: i32, n: i32) -> i32 {
    if x < 0 { 0 } else if x >= n { n - 1 } else { x }
}

fn blur_band(src: &[i32], dst: &mut [i32], n: usize, row_start: usize, row_end: usize) {
    let ni = n as i32;
    for row in row_start..row_end {
        let local_row = row - row_start;
        for col in 0..n {
            let mut acc: i32 = 0;
            for di in -1i32..=1 {
                let nr = clampi(row as i32 + di, ni) as usize;
                for dj in -1i32..=1 {
                    let nc = clampi(col as i32 + dj, ni) as usize;
                    acc += K[((di + 1) * 3 + (dj + 1)) as usize] * src[nr * n + nc];
                }
            }
            dst[local_row * n + col] = acc / 16;
        }
    }
}

fn main() {
    let mut args = env::args().skip(1);
    let cores: usize = args
        .next()
        .and_then(|s| s.parse().ok())
        .unwrap_or(1);
    let n: usize = args
        .next()
        .and_then(|s| s.parse().ok())
        .unwrap_or(256);

    let cores = cores.max(1).min(n);
    let nn = n * n;

    let mut src = vec![0i32; nn];
    let mut dst = vec![0i32; nn];

    let mut s: i64 = 42;
    for v in src.iter_mut() {
        s = lcg(s);
        *v = (s % 256) as i32;
    }

    // Build floor-division row bands (reused each pass).
    let bands: Vec<(usize, usize)> = (0..cores)
        .map(|w| (w * n / cores, (w + 1) * n / cores))
        .collect();

    let t0 = std::time::Instant::now();
    for _pass in 0..PASSES {
        // Split `dst` into per-band mutable slices.
        let mut band_slices: Vec<*mut [i32]> = Vec::with_capacity(cores);
        {
            let mut remaining: &mut [i32] = &mut dst;
            let mut offset = 0usize;
            for w in 0..cores {
                let (_, end) = bands[w];
                let len = (end - offset) * n;
                let (head, tail) = remaining.split_at_mut(len);
                band_slices.push(head as *mut [i32]);
                remaining = tail;
                offset = end;
            }
        }

        // src is read-only for all workers this pass.
        let src_ref: &[i32] = &src;

        std::thread::scope(|scope| {
            for w in 0..cores {
                let (start, end) = bands[w];
                let ptr = band_slices[w];
                // SAFETY: band_slices[w] is a unique non-overlapping segment of `dst`
                // derived from split_at_mut; no other thread touches it.
                let dst_band: &mut [i32] = unsafe { &mut *ptr };
                scope.spawn(move || {
                    blur_band(src_ref, dst_band, n, start, end);
                });
            }
        });
        // All workers have joined; swap buffers (barrier satisfied by scope end).
        std::mem::swap(&mut src, &mut dst);
    }
    eprintln!("COMPUTE_NS {}", t0.elapsed().as_nanos());

    // Serial checksum over `src` (the final buffer after PASSES swaps).
    let mut h: i64 = 0;
    for v in &src {
        h = (h * 31 + *v as i64) % P;
    }
    println!("{}", h);
    println!("blur({})", n);
}
